### PR TITLE
Added telescope observing and prediction to Astro

### PR DESCRIPTION
**Pull request by aruthas**
I took existing astro training and added the option for people who have over 140 Astrology to use a telescope and predict and observe to train Astrology during combat

### CODE REVIEW

File: combat-trainer.lic [hunk 1]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
+      fput('engage')
+	else
+	  retreat
+      check_heavens(game_state)
+      @equipment_manager.wield_weapon(game_state.weapon_name, game_state.weapon_skill)
+      Flags.reset('bad-search')
+	end
+  end
+  
+  def check_heavens(game_state)
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: it uses the same logic as walking astro.  which is a dumbed down version of astrology, which pools from common-arcana

File: combat-trainer.lic [hunk 2]
```diff
@@ -1939,6 +1939,16 @@ class TrainerProcess
 
     @pet_count = @skill_map.include?('Locks') ? get_boxes(@pet_box_source).count : 0
     echo("  @pet_count: #{@pet_count}") if $debug_mode_ct
+	
+	@have_divination_bones = settings.have_divination_bones
+
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: ok will do tomorrow

File: combat-trainer.lic [hunk 3]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
     retreat
-    bput('predict weather', 'You predict that', 'You are far too', 'you lack the skill to grasp them fully')
+	if DRSkill.getrank('Astrology') <= 120
+      bput('predict weather', 'You predict that', 'You are far too', 'you lack the skill to grasp them fully')
+      waitrt?
+      fput('engage')
+	else
+	  retreat
+      check_heavens(game_state)
+      @equipment_manager.wield_weapon(game_state.weapon_name, game_state.weapon_skill)
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: Thanks for looking this over.  I will see what I can do tomorrww

File: combat-trainer.lic [hunk 4]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
+      Flags.reset('bad-search')
+	end
+  end
+  
+  def check_heavens(game_state)
+    @equipment_manager.stow_weapon(game_state.weapon_name)
+    echo "Getting telescope"
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: yep

File: combat-trainer.lic [hunk 5]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
+      return
+    when "You are already holding that."
+      determine_time
+    end
+  end
+  
+  def determine_time
+    echo "Determining time"
+    case DRC.bput("time","it is dawn", "morning", "midday", "noon", "afternoon", "dusk", "sunset", "evening", "midnight", "night", "almost dawn", "approaching sunrise")
+    when "it is dawn"
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: not sure what you mean??

File: combat-trainer.lic (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpherbig) [reply combat-trainer.lic]: This is a whole lot of code to add. Can we reduce duplication in any way?

File: combat-trainer.lic (hunk 5 above)
[the remark this answers is not in the document]
Comment (rpherbig) [reply combat-trainer.lic]: You can make this code a lot more compact

```
    case DRC.bput("time","it is dawn", "morning", "midday", "noon", "afternoon", "dusk", "sunset", "evening", "midnight", "night", "almost dawn", "approaching sunrise")
    when "it is dawn", "morning", ..........
      day_center
    when "sunset", "evening", .............
      night_center
    end
```

File: combat-trainer.lic [hunk 6]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
     return if game_state.retreating?
 
     retreat
-    bput('predict weather', 'You predict that', 'You are far too', 'you lack the skill to grasp them fully')
+	if DRSkill.getrank('Astrology') <= 120
+      bput('predict weather', 'You predict that', 'You are far too', 'you lack the skill to grasp them fully')
+      waitrt?
+      fput('engage')
+	else
```
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: its pretty fast, shorter than hunt

File: combat-trainer.lic (hunk 1 above)
[the remark this answers is not in the document]
Comment (aruthas) [reply combat-trainer.lic]: I could move determine time to common?

[the remark this answers is not in the document]
Comment (jandersson) [reply combat-trainer.lic]: Hi, after checking this, I am creating a common-moonmage module to help support. I'll deduplicate some stuff in `astrology` and `walkingastro`, but keep it pretty simple in the first iteration, just to make sure all the connections are made.

**Comment by jandersson** [comment]
Works well for crannach

File: combat-trainer.lic [hunk 7]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
+      fput("align #{@astrology_prediction_skills_survival}")
+      waitrt?
+      pause 1
+      execute_predict
+    end
   end
 
+  def execute_predict 
+    echo "Predicting"
+    fput("retreat")
```
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: aka change this from fput("retreat") to just retreat

**Comment by Sarvatt** [comment]
Could you add your yaml with the setting so there there is an example to point people to?

**Comment by aruthas** [comment]
Yep, will do needs to be sanitized. Will add it tonight

File: combat-trainer.lic (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: this is fixed in the latest update

File: combat-trainer.lic (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: fixed in the latest update, the debug echos were removed

File: combat-trainer.lic (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: fixed now

File: combat-trainer.lic (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: fixed

File: combat-trainer.lic [hunk 8]
```diff
@@ -2298,11 +2308,204 @@ class TrainerProcess
...
+        echo "Switch to Heart"
+        night_center
+      end
+      if line.include?("You are a bit too busy performing to do that.")
+        echo "Playing an instrument, retrying"
+        fput("stop play")
+        day_center
+      end
+      observe
+      return
```
[the remark this answers is not in the document]
Comment (Sarvatt) [reply combat-trainer.lic]: this is fixed now but the context is far enough away that its not auto updated

File: combat-trainer.lic [hunk 9]
```diff
@@ -2298,9 +2313,156 @@ class TrainerProcess
...
+
+  def execute_predict
+    retreat
+    case DRC.bput('predict future', 'You are far too occupied by present matters to immerse yourself in matters of the future', 'You look inside yourself')
+    when 'You are far too occupied by present matters to immerse yourself in matters of the future'
+      retreat
+      waitrt?
+      execute_predict
+    when 'You look inside yourself'
+      stow_hands
```
Comment (jandersson) [inline combat-trainer.lic]: ❌ DRCI needs to be included to find `stow_hands` otherwise its an error

**Comment by jandersson** [comment]
Updated code throws an error with `stow_hands` but after adding an `include DRCI` to the top of `TrainingProcess` it works just fine! 

Really like this <3

**Comment by jandersson** [comment]
Brainstorming a bit, we can add a future update to get the time, cross reference our astral data and find something to observe that adds more than sun or heart to the prediction pools. We'd need to add more information to the astral data though, like what season, what time of day.

File: combat-trainer.lic (hunk 9 above)
Comment (aruthas) [reply combat-trainer.lic]: Fixed that

**Comment by aruthas** [comment]
Originally when it was conceptualized, we decided on the sun and heart because they were always out, and if you muff a survival prediction, meh.  Not sure if we want to add things like weapons or magic.  Could be dangerous in combat, or armor.

**Comment by aruthas** [comment]
What else do I need to fix??  Would like to get this done before I start school.... Makes me feel like I have contributed.

**Comment by jandersson** [comment]
> Originally when it was conceptualized, we decided on the sun and heart because they were always out, and if you muff a survival prediction, meh. Not sure if we want to add things like weapons or magic. Could be dangerous in combat, or armor.

That makes sense. We can save the high falootin ideas for later.

> What else do I need to fix?? Would like to get this done before I start school.... Makes me feel like I have contributed.

Nothing that im aware of. Ive been testing it for the last week, the only issue I ran into was the `stow_hands` one, which is fixed.

**Comment by Sarvatt** [comment]
Please pull this, its been tested way more than any other change by people who care. Moon mages are weird. What reason do you have not to pull it?

**Comment by rpherbig** [comment]
> Please pull this, its been tested way more than any other change by people who care. Moon mages are weird. What reason do you have not to pull it?

Sorry, the holidays have been crazy.

**Pull request merged**